pkgs/misc/vscode-extensions/default.nix

```diff
@@ -144,6 +144,23 @@ let
         };
       };
 
+      apollographql.vscode-apollo = buildVscodeMarketplaceExtension {
+        mktplcRef = {
+          name = "vscode-apollo";
+          publisher = "apollographql";
+          version = "1.19.9";
+          sha256 = "sha256-iJpzNKcuQrfq4Z0LXuadt6OKXelBbDQg/vuc7NJ2I5o=";
```

Comment:
I reversed engineer my way to find the `sha256` value. What is the easiest method to get this hash?

Reply:
[TOFU](https://en.wikipedia.org/wiki/Trust_on_first_use) I think, that is to supply the fixed-output derivation with a probably-wrong output hash, and force the build to fail with a hash mismatch error, which contains in its error message the actual output hash. You may use tools like [nix-prefetch](https://github.com/msteen/nix-prefetch) and [nix-universal-prefetch](https://github.com/samueldr/nix-universal-prefetch) to do this.

Reply:
Here is how I did it after understanding that I had to base64 encode the SHA256 in binary form (not base64 the hex representation of the SHA256).

1. Go on https://marketplace.visualstudio.com/items?itemName=apollographql.vscode-apollo
2. Download the extension by clicking the "Download Extension" link (cannot get the url of the resource since the ling is not a HTML `<a>` element)
3. Run `sha256sum apollographql.vscode-apollo-1.19.9.vsix | cut -f1 -d\ | xxd -r -p | base64 | awk '{print "sha256-" $0}'`

As indicated in 2, the resource URL is not available so I do not see how to use the `nix-prefetch`.

Perhaps the existing vscode-extensions maintainers (@ratsclub @ImGabe @SuperSandro2000) could indicate how they generate the `sha256` value? And why they use the type of `sha256` value starting with "sha256-" as opposed to the type of hash found in https://github.com/NixOS/nixpkgs/blob/ff32943261542f1d47c3f4e09abeb0a3b7d9adb4/pkgs/misc/vscode-extensions/default.nix#L550 for example?

Reply:
SRI representation is the default for Nix 2.4 (or later) I think, but actually all kinds of encodings are accepted, you may check https://nixos.wiki/wiki/Nix_Hash for details.

Actually the url is https://apollographql.gallery.vsassets.io/_apis/public/gallery/publisher/apollographql/extension/vscode-apollo/1.19.9/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage.

And also you can also just fill sha256 with lib.fakeSha256 (`0000000000000000000000000000000000000000000000000000`) and try to build the extension with `nix-build -A vscode-extensions.apollographql.vscode-apollo` then copy the correct hash from the error.

Reply:
I mostly do what @bobby285271 said: `sha256 = lib.fakeSha256` and try to build. However, something to keep in mind is to change the version attribute when you update the extension, sometimes I forget about it and wonder why it didn't work :smile: